### PR TITLE
Add wizard class as an argument to the onCompleteAction

### PR DESCRIPTION
**Pull request by jorostoyanov**
The following PR will allow the `onCompleteAction` class to have information about the wizard in addition to the already available payload.

E.g.

```
<?php

namespace App\Wizards\Registration;

use Arcanist\Action\WizardAction;
use Arcanist\Action\ActionResult;

class RegistrationAction extends WizardAction
{
    public function execute($payload, $wizard): ActionResult
    {
        return $this->success();
    }
}
```

### CODE REVIEW

**Comment by MillerAdulu** [comment]
Interesting PR @jorostoyanov,

Is there a particular use case where this would be needed that can't be handled by the package as is at the moment?

**Comment by jorostoyanov** [comment]
Yes!

Imagine you have a quiz and want to insert all questions and their respective answers for each user in the database as separated records (e.g. a QuizAnswer Model). Currently the library stores all answers as JSON column `data`. This makes it hard to later do any bulk operations on the answers (e.g. show stats for all submitted answers).

Currently, there is no way to get extra information about the wizard such as ID, the steps rules and so on. The only way to get that information is to get the wizard ID from the Request class and instantiate the wizard again.

However, all of the needed information is available in the `$wizard` argument.

Hope this makes sense! :-)

**Comment by MillerAdulu** [comment]
Makes sense @jorostoyanov,

In regards to a few points that you have mentioned, 

> Currently, there is no way to get extra information about the wizard such as ID

In the class that extends the Wizard, you can always override the `transformWizardData` data method to insert whatever data you want so that it's available in the payload for the action class. Here's how I did it a while back. 

```php
protected function transformWizardData(): mixed
    {
        return array_merge(
            ['wizard_id' => $this->id,],
            $this->data(),
        );
    }
```
This way you are able to pass in the ID of the Wizard should you want to associate a specific user with a wizard. 

Also, a point to note, the wizard_id shouldn't be used as a foreign key at all because after a Wizard is successfully completed, it's deleted. So the ID can only be good for marking sessions in case a user can submit without logging in multiple times and you don't want them to have multiple records without a unique identifier for each. 


In regard to: 
> getting data about the steps, rules and so on

In my opinion, I am not entirely sure of why those would need to be persisted. The rules are made in code and as long as the code exists, the rules are still intact. Thus properly defining the rules and steps would result in you needing all that you need without having to go to the extent of persisting the steps.

In regard to:

> Imagine you have a quiz and want to insert all questions and their respective answers for each user in the database as separated

Assuming that you have 10 questions, each with its own step, you could try something like (my opinion),
- Have a question_id as part of the `viewData` method
- Then you have a hidden field as part of the form and defined in the `fields()` definition in the step itself. So that you have `question_id` and `answer` for each step. 
- This will then allow you to have them as part of the payload during the wizard action and you can iterate over the question ID and insert the models however you want without necessarily having to pass the entire wizard to the transform action. 


In conclusion, I seem to get the indication that you would like to use the wizards table for a lot of the business logic yet I see that the package author may have intended it to be useful only for wizards and after submission, the wizard record should no longer exist.

**Comment by jorostoyanov** [comment]
Makes sense all that you've described @MillerAdulu 

No, I am not intending to use the `wizards` table for business logic, as I am creating a separate `submissions` table to hold all of the information.

Getting the wizard ID was just an example of what can be used in the action class. Probably, mentioning the rules was not the best example as, like you've said, it is related to the validation logic at the moment of step submission, and it isn't something that is tightly connected to the submission record itself.

I just wanted to showcase that in the action class sometimes you might need the whole wizard setup class data (e.g. the steps structure with the questions, etc). Then use them as per need. Since the action class is the last step before completing the wizard and is tightly connected to it, I think the more information you have, the better.

Let the developer decide how and whether to use the data :-)

**Comment by ksassnowski** [comment]
Thanks for the contribution and sorry for the late response!

Passing the wizard to every action seems like a rather invasive change for what I think is a bit of an unusual use-case. Changing the method signature would be a breaking change, even though that wouldn't be as big of a deal since we’re still pre 1.0.

The bigger problem I see with that is that it would now couple the action class very tightly to the wizard class itself. This means that in any test I write for the action, I would have to instantiate the wizard class as well. The wizard class is fairly complicated to instantiate so I’d like to avoid this.

Like @MillerAdulu mentioned, the `transformWizardData` method is what you should use for your use-case. If you need the wizard in your action, you can add it to the action’s payload like so

```php
// In your wizard
public function transformWizardData(): mixed
{
    return array_merge(['wizard' => $this], $this->data());
}

// In your action
public function execute($payload): ActionResult
{
     // Access the wizard instance like this.
     $wizard = $payload['wizard'];
}
```
